File: 05.Grafos/adjmatrixedgeweighteddigraph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include <iomanip>
#include <stdexcept>
#include "adjmatrixedgeweighteddigraph.h"

AdjMatrixEdgeWeightedDigraph::AdjMatrixEdgeWeightedDigraph(EdgeWeightedDigraph& graph) {

    g = graph;
    totalVertices = 0;
    for (const auto& v : g.getVerts()) {
        dic[v] = totalVertices;
        dic2[totalVertices] = v;
        totalVertices++;
    }

    adj = (int**) malloc(totalVertices * sizeof(int*));
    data = (int*) malloc(totalVertices * totalVertices * sizeof(int));

    for(int i=0; i<totalVertices; i++) {
      adj[i] = &data[i*totalVertices];
    }

    for(int i=0; i<totalVertices; i++)
      for(int j=0; j<totalVertices; j++)
        adj[i][j] = -1; // Initialize all weights to -1 (no edge)

    for (const auto& e : g.getEdges()) {
        addEdge(e);
    }
}

int AdjMatrixEdgeWeightedDigraph::mapToArray(const std::string& v) const {
    return dic.at(v);
}

std::string AdjMatrixEdgeWeightedDigraph::mapToString(int v) const {
    return dic2.at(v);
}

int AdjMatrixEdgeWeightedDigraph::getTotalVertices() const {
    return totalVertices;
}

int AdjMatrixEdgeWeightedDigraph::getTotalEdges() const {
    return totalEdges;
}

void AdjMatrixEdgeWeightedDigraph::addEdge(const Edge& e) {
    std::string v = e.v;
    std::string w = e.w;
    int v1 = dic[v];
    int w1 = dic[w];
    if (adj[v1][w1] == -1) {
        adj[v1][w1] = e.weight;
        totalEdges++;
    }
}

std::vector<Edge> AdjMatrixEdgeWeightedDigraph::getAdj(const std::string& v) {
    return g.getAdj(v);
}

std::string AdjMatrixEdgeWeightedDigraph::toString() const {
    std::ostringstream oss;
    oss << "  ";
    for (int i = 0; i < totalVertices; ++i) {
        oss << std::setw(6) << i;
    }
    oss << "\n";
    for (int i = 0; i < totalVertices; ++i) {
        oss << i << " ";
        for (int j = 0; j < totalVertices; ++j) {
            if (adj[i][j] != -1) {
                oss << std::setw(5) << std::fixed << std::setprecision(2) << adj[i][j] << " ";
            } else {
                oss << "----- ";
            }
        }
        oss << "\n";
    }
    return oss.str();
}

AdjMatrixEdgeWeightedDigraph::~AdjMatrixEdgeWeightedDigraph() {
    free(data);
    free(adj);
}

File: 05.Grafos/adjmatrixedgeweighteddigraph.py
from edgeweighteddigraph import EdgeWeightedDigraph

class AdjMatrixEdgeWeightedDigraph:
    def __init__(self, g):
        self.g = g

        self.dic = {}
        self.dic2 = {}
        self.total_vertices = len(g.getVerts())
        for idx, v in enumerate(sorted(g.getVerts())):
            self.dic[v] = idx
            self.dic2[idx] = v

        self.total_edges = 0
        self.adj = [[None for _ in range(self.total_vertices)] for _ in range(self.total_vertices)]
        for e in g.getEdges():
            self.addEdge(e)

    def mapToArray(self, v):
        return self.dic[v]

    def mapToString(self, v):
        return self.dic2[v]

    def getTotalEdges(self):
        return self.total_edges

    def addEdge(self, e):
        v = e[0]
        w = e[1]
        weight = e[2]
        v1 = self.dic[v]
        w1 = self.dic[w]
        if self.adj[v1][w1] is None:
            self.total_edges += 1
            self.adj[v1][w1] = weight

    def getAdj(self, v):
        return self.g.getAdj(v)

    def toDot(self):
        return self.g.toDot()

    def __str__(self):
        lines = []
        header = "  " + " ".join(f"{i:<5}" for i in range(self.total_vertices))
        lines.append(header)
        for i in range(self.total_vertices):
            row = [f"{i}"]
            for j in range(self.total_vertices):
                weight = self.adj[i][j]
                if weight:
                    row.append(f"{weight:5.2f}")
                else:
                    row.append("-----")
            lines.append(" ".join(row))
        return "\n".join(lines)


# Exemplo de uso:
if __name__ == "__main__":
    g = EdgeWeightedDigraph("exemplos/tinyEWD.txt")
    ag = AdjMatrixEdgeWeightedDigraph(g)
    print(ag.toDot())
    print(ag)

